File: scripts/quant_api_client.js
#!/usr/bin/env node
/**
 * Official Hexagon Quant Terminal CLI & API Tool (2026)
 * Communicates with the Rust REST API Server (http://127.0.0.1:5000)
 * Supports querying, live inspection, and JSON exports for backtests, trades, monte-carlo, and EDA.
 */

const fs = require('fs');
const path = require('path');
const http = require('http');

const DEFAULT_API_BASE = process.env.QUANT_API_BASE || 'http://127.0.0.1:5000/api';

function requestApi(endpoint, apiBase = DEFAULT_API_BASE) {
  return new Promise((resolve, reject) => {
    const url = `${apiBase.replace(/\/+$/, '')}/${endpoint.replace(/^\/+/, '')}`;
    http.get(url, (res) => {
      let raw = '';
      res.on('data', chunk => raw += chunk);
      res.on('end', () => {
        try {
          if (res.statusCode >= 400) {
            reject(new Error(`HTTP ${res.statusCode}: ${raw}`));
          } else {
            resolve(JSON.parse(raw));
          }
        } catch (e) {
          reject(new Error(`Failed to parse JSON response: ${e.message}`));
        }
      });
    }).on('error', (err) => {
      reject(new Error(`Gagal menghubungi API (${url}): ${err.message}. Pastikan backend Rust berjalan (./dev.sh).`));
    });
  });
}

function exportJson(data, outputPath) {
  const absPath = path.resolve(process.cwd(), outputPath);
  fs.mkdirSync(path.dirname(absPath), { recursive: true });
  fs.writeFileSync(absPath, JSON.stringify(data, null, 2), 'utf-8');
  const size = fs.statSync(absPath).size;
  console.log(`✅ Berhasil mengekspor JSON ke: ${outputPath} (${size} bytes)`);
}

async function main() {
  const args = process.argv.slice(2);
  const cmd = args[0];

  if (!cmd || cmd === '--help' || cmd === '-h') {
    console.log(`
╔═════════════════════════════════════════════════════════════════════════╗
║   ⚡ HEXAGON QUANT TERMINAL • API & BACKTEST DATA EXPORTER (2026)      ║
╚═════════════════════════════════════════════════════════════════════════╝

Penggunaan:
  node scripts/quant_api_client.js <command> [options]

Perintah:
  health                     Cek status health REST API server
  backtest [--export FILE]   Ambil ringkasan portfolio backtest 10-tahun
  detailed <PAIR> [--export] Ambil detailed backtest report per pair (EURGBP, EURUSD, ...)
  trades <PAIR> [--export]   Ambil daftar settled trades per pair
  monte-carlo <PAIR>         Ambil simulasi Monte Carlo 1.000 path
  scorecard                  Ambil 7-Pilar scorecard & revenue share status
  sync <PAIR>                Trigger delta sync check
  dump-all [--out-dir DIR]   Ekspor seluruh dataset backtest & scorecard ke folder JSON

Contoh:
  node scripts/quant_api_client.js backtest
  node scripts/quant_api_client.js detailed EURGBP --export reports/EURGBP_backtest.json
  node scripts/quant_api_client.js trades EURGBP --export reports/EURGBP_trades.json
  node scripts/quant_api_client.js dump-all --out-dir reports/backtest
`);
    process.exit(0);
  }

  const exportIdx = args.indexOf('--export');
  const exportPath = exportIdx !== -1 ? args[exportIdx + 1] : null;
  const isJsonOnly = args.includes('--json');

  try {
    if (cmd === 'health') {
      const data = await requestApi('health');
      if (exportPath) exportJson(data, exportPath);
      else if (isJsonOnly) console.log(JSON.stringify(data, null, 2));
      else {
        console.log(`🟢 API STATUS: ${data.status.toUpperCase()} | Version: ${data.version} | Engine: ${data.engine}`);
      }
    } else if (cmd === 'backtest') {
      const data = await requestApi('backtest');
      if (exportPath) exportJson(data, exportPath);
      else if (isJsonOnly) console.log(JSON.stringify(data, null, 2));
      else {
        console.log('═════════════════════════════════════════════════════════════════════════');
        console.log('📊 PORTFOLIO MULTI-PAIR BACKTEST SUMMARY (DUKASCOPY 10-YEAR DATASET)');
        console.log('═════════════════════════════════════════════════════════════════════════');
        console.log(`Total Valued Pips   : ${data.total_valued_pips > 0 ? '+' : ''}${data.total_valued_pips.toFixed(1)} VP`);
        console.log(`Portfolio Win Rate  : ${data.portfolio_win_rate_pct.toFixed(1)}%`);
        console.log(`WFER OOS Stability  : ${data.wfer_pct.toFixed(1)}%`);
        console.log(`TF Reward Qualified : ${data.is_tf_qualified ? '✅ YES (>= 300 VP)' : '❌ IN PROGRESS'}`);
        console.log('─────────────────────────────────────────────────────────────────────────');
        console.log(`${'PAIR'.padEnd(10)} ${'TRADES'.padEnd(8)} ${'WIN%'.padEnd(8)} ${'RAW PIPS'.padEnd(12)} ${'VALUED PIPS'.padEnd(14)} ${'PF'.padEnd(6)} STATUS`);
        console.log('─────────────────────────────────────────────────────────────────────────');
        for (const r of data.reports || []) {
          const status = r.is_tf_qualified ? '✅ TF OK' : '⚠️ REVIEW';
          console.log(`${r.symbol.padEnd(10)} ${String(r.total_trades).padEnd(8)} ${(r.win_rate_percent.toFixed(1) + '%').padEnd(8)} ${(r.total_raw_pips.toFixed(1)).padEnd(12)} ${(r.total_valued_pips.toFixed(1)).padEnd(14)} ${(r.profit_factor.toFixed(2)).padEnd(6)} ${status}`);
        }
        console.log('═════════════════════════════════════════════════════════════════════════');
      }
    } else if (cmd === 'detailed') {
      const sym = (args[1] || 'EURGBP').toUpperCase();
      const data = await requestApi(`backtest/detailed/${sym}`);
      if (exportPath) exportJson(data, exportPath);
      else if (isJsonOnly) console.log(JSON.stringify(data, null, 2));
      else {
        const rep = data.report || data;
        console.log('═════════════════════════════════════════════════════════════════════════');
        console.log(`📈 DETAILED BACKTEST REPORT: ${sym} (H1 Candlestick • 10-Year Dukascopy)`);
        console.log('═════════════════════════════════════════════════════════════════════════');
        console.log(`Total Trades        : ${rep.total_trades}`);
        console.log(`Win Rate            : ${(rep.win_rate_percent || rep.win_rate_pct || 0).toFixed(1)}% (${rep.winning_trades}W / ${rep.losing_trades}L)`);
        console.log(`Total Raw PnL       : ${(rep.total_raw_pips || 0) > 0 ? '+' : ''}${(rep.total_raw_pips || 0).toFixed(1)} pips`);
        console.log(`Total Valued Pips   : ${(rep.total_valued_pips || 0) > 0 ? '+' : ''}${(rep.total_valued_pips || 0).toFixed(1)} VP`);
        console.log(`Profit Factor       : ${(rep.profit_factor || 0).toFixed(2)}`);
        console.log(`Recovery Factor     : ${(rep.recovery_factor || 0).toFixed(2)}`);
        console.log(`Max Drawdown (Pips) : ${(rep.max_drawdown_pips || 0).toFixed(1)} pips`);
        console.log(`TF Reward Qualified : ${rep.is_tf_qualified ? '✅ YES (>= 300 VP)' : '❌ IN PROGRESS'}`);
        console.log(`Settled Trades List : ${data.trades?.length || rep.trades?.length || 0} trades recorded`);
        console.log('═════════════════════════════════════════════════════════════════════════');
      }
    } else if (cmd === 'trades') {
      const sym = (args[1] || 'EURGBP').toUpperCase();
      const data = await requestApi(`backtest/trades/${sym}`);
      if (exportPath) exportJson(data, exportPath);
      else if (isJsonOnly) console.log(JSON.stringify(data, null, 2));
      else {
        console.log('═════════════════════════════════════════════════════════════════════════');
        console.log(`📜 LIST OF SETTLED TRADES: ${sym} (Total: ${data.length} Trades)`);
        console.log('═════════════════════════════════════════════════════════════════════════');
        console.log(`${'#'.padEnd(4)} ${'ACTION'.padEnd(12)} ${'ENTRY'.padEnd(10)} ${'EXIT'.padEnd(10)} ${'PNL PIPS'.padEnd(10)} ${'VALUED PIPS'.padEnd(12)} RESULT`);
        console.log('─────────────────────────────────────────────────────────────────────────');
        const showTrades = data.slice(0, 30);
        showTrades.forEach((t, i) => {
          const res = t.pnl_pips > 0 ? '🟢 WIN' : '🔴 LOSS';
          console.log(`${String(i + 1).padEnd(4)} ${t.action.padEnd(12)} ${t.entry_price.toFixed(5).padEnd(10)} ${t.exit_price.toFixed(5).padEnd(10)} ${(t.pnl_pips > 0 ? '+' : '') + t.pnl_pips.toFixed(1).padEnd(8)} ${(t.valued_pips > 0 ? '+' : '') + t.valued_pips.toFixed(1).padEnd(10)} ${res}`);
        });
        if (data.length > 30) {
          console.log(`... dan ${data.length - 30} trade lainnya (gunakan --export file.json untuk dump semua).`);
        }
        console.log('═════════════════════════════════════════════════════════════════════════');
      }
    } else if (cmd === 'monte-carlo') {
      const sym = (args[1] || 'EURGBP').toUpperCase();
      const data = await requestApi(`monte-carlo/${sym}`);
      if (exportPath) exportJson(data, exportPath);
      else if (isJsonOnly) console.log(JSON.stringify(data, null, 2));
      else {
        console.log('═════════════════════════════════════════════════════════════════════════');
        console.log(`🎲 MONTE CARLO 1,000-PATH RESAMPLING: ${sym}`);
        console.log('═════════════════════════════════════════════════════════════════════════');
        console.log(`Iterations          : ${data.iterations} paths`);
        console.log(`Risk of Ruin (DD>20): ${data.risk_of_ruin_pct.toFixed(2)}% (Zero Ruin Guarantee)`);
        console.log(`Worst-Case Max DD   : -${data.worst_case_max_dd_pct.toFixed(1)} VP (5th Percentile Floor)`);
        console.log(`Median Expected VP  : +${data.median_ending_vp.toFixed(1)} VP (50th Percentile)`);
        console.log(`95% Confidence Int  : [${data.confidence_interval_95[0].toFixed(1)}, ${data.confidence_interval_95[1].toFixed(1)}] VP`);
        console.log('═════════════════════════════════════════════════════════════════════════');
      }
    } else if (cmd === 'scorecard') {
      const data = await requestApi('scorecard');
      if (exportPath) exportJson(data, exportPath);
      else if (isJsonOnly) console.log(JSON.stringify(data, null, 2));
      else {
        console.log('═════════════════════════════════════════════════════════════════════════');
        console.log('🏆 TRADERS FAMILY 7-PILLAR SCORECARD & REVENUE SHARE ELIGIBILITY');
        console.log('═════════════════════════════════════════════════════════════════════════');
        console.log(`Total Score         : ${data.total_score} / 28 (${(data.total_score/28*100).toFixed(1)}%)`);
        console.log(`Revenue Share Tier  : ${data.revenue_share_tier}`);
        console.log(`Max Revenue Share   : ${data.max_revenue_share_pct}%`);
        console.log('═════════════════════════════════════════════════════════════════════════');
      }
    } else if (cmd === 'audit') {
      const sym = (args[1] || 'EURGBP').toUpperCase();
      const data = await requestApi(`audit/pair/${sym}`);
      if (exportPath) exportJson(data, exportPath);
      else if (isJsonOnly) console.log(JSON.stringify(data, null, 2));
      else {
        console.log('═════════════════════════════════════════════════════════════════════════');
        console.log(`🔬 COMPLETE 360° QUANT AUDIT REPORT: ${sym}`);
        console.log('═════════════════════════════════════════════════════════════════════════');
        console.log(`Pair Tier & Multiplier: Tier ${data.tier} (${data.multiplier}x VP Multiplier)`);
        console.log(`Total Trades          : ${data.total_trades} trades (${data.winning_trades} Win / ${data.losing_trades} Loss)`);
        console.log(`Win Rate              : ${data.win_rate_pct.toFixed(1)}%`);
        console.log(`Raw PnL & Valued Pips : ${data.total_raw_pips > 0 ? '+' : ''}${data.total_raw_pips.toFixed(1)} pips | ${data.total_valued_pips > 0 ? '+' : ''}${data.total_valued_pips.toFixed(1)} VP`);
        console.log(`Profit & Rec. Factor  : PF: ${data.profit_factor.toFixed(2)} | RecF: ${data.recovery_factor.toFixed(2)}`);
        console.log(`Sharpe / Sortino / Cal: ${data.sharpe_ratio.toFixed(2)} / ${data.sortino_ratio.toFixed(2)} / ${data.calmar_ratio.toFixed(2)}`);
        console.log(`Max Drawdown          : ${data.max_drawdown_pips.toFixed(1)} pips (${data.max_drawdown_vp.toFixed(1)} VP)`);
        console.log(`TF Reward Qualified   : ${data.is_tf_qualified ? '✅ YES (>= 300 VP)' : '❌ IN PROGRESS'}`);
        console.log(`Data Footprint        : ${data.provenance.total_bars.toLocaleString()} Bars (${data.provenance.provider_name})`);
        console.log(`Math Integrity Check  : ${data.provenance.mathematical_integrity_pct.toFixed(1)}% Passed (Zero Invariant Errors)`);
        console.log(`Monthly History       : ${data.monthly_breakdown.length} months recorded`);
        console.log(`Equity Curve Points   : ${data.equity_curve.length} points`);
        console.log(`Settled Trades Log    : ${data.trades.length} individual trade records`);
        console.log('═════════════════════════════════════════════════════════════════════════');
      }
    } else if (cmd === 'audit-full') {
      const data = await requestApi('audit/full');
      if (exportPath) exportJson(data, exportPath);
      else if (isJsonOnly) console.log(JSON.stringify(data, null, 2));
      else {
        console.log('═════════════════════════════════════════════════════════════════════════');
        console.log('🏛️ FULL PORTFOLIO 360° QUANT AUDIT REPORT (UI-MIRRORED EXACT NUMBERS)');
        console.log('═════════════════════════════════════════════════════════════════════════');
        console.log(`Generated At          : ${data.generated_at}`);
        console.log(`Portfolio Valued Pips : ${data.total_portfolio_valued_pips > 0 ? '+' : ''}${data.total_portfolio_valued_pips.toFixed(1)} VP / ${data.monthly_tf_target_vp} VP Target`);
        console.log(`Portfolio Status      : ${data.is_portfolio_tf_qualified ? '✅ QUALIFIED REVENUE SHARE' : '❌ IN PROGRESS'}`);
        console.log(`Portfolio Win Rate    : ${data.portfolio_win_rate_pct.toFixed(1)}% (${data.total_portfolio_trades} Total Trades)`);
        console.log(`7-Pillar Scorecard    : ${data.scorecard.total_score} / ${data.scorecard.max_score} (${data.scorecard.score_pct.toFixed(1)}% - ${data.scorecard.revenue_share_tier})`);
        console.log(`Walk-Forward WFER     : ${data.walk_forward.wfer_pct.toFixed(1)}% (${data.walk_forward.total_verified_bars.toLocaleString()} Real Dukascopy Bars)`);
        console.log('─────────────────────────────────────────────────────────────────────────');
        console.log('PER-PAIR AUDIT BREAKDOWN:');
        for (const p of data.pairs) {
          console.log(`  • ${p.symbol.base}/${p.symbol.quote} (Tier ${p.tier}): ${p.total_trades} trades | Win: ${p.win_rate_pct.toFixed(1)}% | VP: ${p.total_valued_pips > 0 ? '+' : ''}${p.total_valued_pips.toFixed(1)} VP | PF: ${p.profit_factor.toFixed(2)} | Bars: ${p.provenance.total_bars.toLocaleString()}`);
        }
        console.log('═════════════════════════════════════════════════════════════════════════');
      }
    } else if (cmd === 'dump-all') {
      const outDirIdx = args.indexOf('--out-dir');
      const outDir = outDirIdx !== -1 ? args[outDirIdx + 1] : 'reports/backtest';
      console.log(`🚀 Memulai dump seluruh dataset kuantitatif ke folder: ${outDir}/ ...\n`);
      
      const bt = await requestApi('backtest');
      exportJson(bt, path.join(outDir, 'portfolio_backtest_summary.json'));
      
      const sc = await requestApi('scorecard');
      exportJson(sc, path.join(outDir, 'scorecard_7_pillar.json'));

      const strats = await requestApi('strategies');
      exportJson(strats, path.join(outDir, 'strategies_catalog.json'));

      const pairs = ['EURGBP', 'USDCHF', 'GBPUSD', 'EURUSD', 'NZDUSD', 'AUDUSD'];
      for (const p of pairs) {
        console.log(`📦 Mengambil data lengkap untuk pair: ${p}...`);
        try {
          const det = await requestApi(`backtest/detailed/${p}`);
          exportJson(det, path.join(outDir, `${p}_detailed_backtest.json`));
        } catch (e) {
          console.warn(`   ⚠️ Detailed backtest ${p}: ${e.message}`);
        }
        try {
          const tr = await requestApi(`backtest/trades/${p}`);
          exportJson(tr, path.join(outDir, `${p}_trades_list.json`));
        } catch (e) {
          console.warn(`   ⚠️ Trades ${p}: ${e.message}`);
        }
        try {
          const mc = await requestApi(`monte-carlo/${p}`);
          exportJson(mc, path.join(outDir, `${p}_monte_carlo.json`));
        } catch (e) {
          console.warn(`   ⚠️ Monte Carlo ${p}: ${e.message}`);
        }
        try {
          const eda = await requestApi(`eda/${p}`);
          exportJson(eda, path.join(outDir, `${p}_eda_health.json`));
        } catch (e) {
          console.warn(`   ⚠️ EDA ${p}: ${e.message}`);
        }
      }

      console.log(`\n🎉 SELESAI! Seluruh data JSON berhasil diekspor ke: ${outDir}/`);
    } else {
      console.error(`❌ Perintah tidak dikenal: ${cmd}. Ketik 'node scripts/quant_api_client.js --help' untuk bantuan.`);
    }
  } catch (err) {
    console.error(`❌ Terjadi error: ${err.message}`);
    process.exit(1);
  }
}

main();

File: scripts/quant_api_client.py
#!/usr/bin/env python3
"""
Official Hexagon Quant Terminal CLI & API Tool (2026)
Communicates with the Rust REST API Server (http://127.0.0.1:5000)
Supports querying, live inspection, and JSON exports for backtests, trades, monte-carlo, and EDA.
"""

import os
import sys
import json
import argparse
import urllib.request
import urllib.error

DEFAULT_API_BASE = os.environ.get("QUANT_API_BASE", "http://127.0.0.1:5000/api")

def request_api(endpoint: str, api_base: str = DEFAULT_API_BASE):
    url = f"{api_base.rstrip('/')}/{endpoint.lstrip('/')}"
    req = urllib.request.Request(
        url,
        headers={"User-Agent": "Hexagon-Quant-CLI/2026", "Accept": "application/json"}
    )
    try:
        with urllib.request.urlopen(req, timeout=10) as resp:
            data = resp.read().decode('utf-8')
            return json.loads(data)
    except urllib.error.URLError as e:
        print(f"❌ Error connecting to API ({url}): {e}", file=sys.stderr)
        print("💡 Pastikan backend Rust berjalan (./dev.sh atau cargo run -p api-server).", file=sys.stderr)
        sys.exit(1)

def export_json(data, output_path: str):
    os.makedirs(os.path.dirname(os.path.abspath(output_path)), exist_ok=True)
    with open(output_path, 'w', encoding='utf-8') as f:
        json.dump(data, f, indent=2)
    print(f"✅ Berhasil mengekspor JSON ke: {output_path} ({os.path.getsize(output_path)} bytes)")

def cmd_health(args):
    data = request_api("health", args.api_base)
    if args.json or args.export:
        if args.export:
            export_json(data, args.export)
        else:
            print(json.dumps(data, indent=2))
    else:
        print("═══════════════════════════════════════════════════════")
        print(f"🟢 API STATUS: {data.get('status', 'OK').upper()}")
        print(f"   Version : {data.get('version', '0.1.0')}")
        print(f"   Engine  : {data.get('engine', 'Rust Axum Hexagonal')}")
        print("═══════════════════════════════════════════════════════")

def cmd_backtest(args):
    data = request_api("backtest", args.api_base)
    if args.export:
        export_json(data, args.export)
    elif args.json:
        print(json.dumps(data, indent=2))
    else:
        print("═════════════════════════════════════════════════════════════════════════")
        print("📊 PORTFOLIO MULTI-PAIR BACKTEST SUMMARY (DUKASCOPY 10-YEAR DATASET)")
        print("═════════════════════════════════════════════════════════════════════════")
        print(f"Total Valued Pips   : {data.get('total_valued_pips', 0):+.1f} VP")
        print(f"Portfolio Win Rate  : {data.get('portfolio_win_rate_pct', 0):.1f}%")
        print(f"WFER OOS Stability  : {data.get('wfer_pct', 0):.1f}%")
        print(f"TF Reward Qualified : {'✅ YES (>= 300 VP)' if data.get('is_tf_qualified') else '❌ IN PROGRESS'}")
        print("─────────────────────────────────────────────────────────────────────────")
        print(f"{'PAIR':<10} {'TRADES':<8} {'WIN%':<8} {'RAW PIPS':<12} {'VALUED PIPS':<14} {'PF':<6} {'STATUS'}")
        print("─────────────────────────────────────────────────────────────────────────")
        for r in data.get('reports', []):
            status = '✅ TF OK' if r.get('is_tf_qualified') else '⚠️ REVIEW'
            print(f"{r.get('symbol'):<10} {r.get('total_trades'):<8} {r.get('win_rate_percent', 0):<8.1f} {r.get('total_raw_pips', 0):<12.1f} {r.get('total_valued_pips', 0):<14.1f} {r.get('profit_factor', 0):<6.2f} {status}")
        print("═════════════════════════════════════════════════════════════════════════")

def cmd_detailed(args):
    symbol = args.symbol.upper()
    data = request_api(f"backtest/detailed/{symbol}", args.api_base)
    if args.export:
        export_json(data, args.export)
    elif args.json:
        print(json.dumps(data, indent=2))
    else:
        print(f"═════════════════════════════════════════════════════════════════════════")
        print(f"📈 DETAILED BACKTEST REPORT: {symbol} (H1 Candlestick)")
        print(f"═════════════════════════════════════════════════════════════════════════")
        print(f"Total Trades        : {data.get('total_trades', 0)}")
        print(f"Win Rate            : {data.get('win_rate_pct', 0):.1f}%")
        print(f"Total Raw PnL       : {data.get('total_raw_pips', 0):+.1f} pips")
        print(f"Total Valued Pips   : {data.get('total_valued_pips', 0):+.1f} VP")
        print(f"Profit Factor       : {data.get('profit_factor', 0):.2f}")
        print(f"Recovery Factor     : {data.get('recovery_factor', 0):.2f}")
        print(f"Max Drawdown (Pips) : {data.get('max_drawdown_pips', 0):.1f} pips")
        print(f"WFER OOS Stability  : {data.get('wfer_pct', 0):.1f}%")
        print(f"Traders in List     : {len(data.get('trades', []))} settled trades")
        print(f"═════════════════════════════════════════════════════════════════════════")

def cmd_trades(args):
    symbol = args.symbol.upper()
    data = request_api(f"backtest/trades/{symbol}", args.api_base)
    if args.export:
        export_json(data, args.export)
    elif args.json:
        print(json.dumps(data, indent=2))
    else:
        print(f"═════════════════════════════════════════════════════════════════════════")
        print(f"📜 LIST OF SETTLED TRADES: {symbol} (Total: {len(data)} Trades)")
        print(f"═════════════════════════════════════════════════════════════════════════")
        print(f"{'#':<4} {'ACTION':<10} {'ENTRY':<10} {'EXIT':<10} {'PNL PIPS':<10} {'VALUED PIPS':<12} {'RESULT'}")
        print("─────────────────────────────────────────────────────────────────────────")
        for i, t in enumerate(data[: args.limit if args.limit else 50], 1):
            res_icon = '🟢 WIN' if t.get('pnl_pips', 0) > 0 else '🔴 LOSS'
            print(f"{i:<4} {t.get('action', 'BUY_LIMIT'):<10} {t.get('entry_price', 0):<10.5f} {t.get('exit_price', 0):<10.5f} {t.get('pnl_pips', 0):<+10.1f} {t.get('valued_pips', 0):<+12.1f} {res_icon}")
        if len(data) > 50 and not args.limit:
            print(f"... dan {len(data) - 50} trade lainnya (gunakan --limit 500 atau --export file.json)")
        print("═════════════════════════════════════════════════════════════════════════")

def cmd_monte_carlo(args):
    symbol = args.symbol.upper()
    data = request_api(f"monte-carlo/{symbol}", args.api_base)
    if args.export:
        export_json(data, args.export)
    elif args.json:
        print(json.dumps(data, indent=2))
    else:
        print(f"═════════════════════════════════════════════════════════════════════════")
        print(f"🎲 MONTE CARLO 1,000-PATH SIMULATION: {symbol}")
        print(f"═════════════════════════════════════════════════════════════════════════")
        print(f"Iterations          : {data.get('iterations', 1000)} permutations")
        print(f"Risk of Ruin        : {data.get('risk_of_ruin_pct', 0):.2f}%")
        print(f"Worst-Case Max DD   : {data.get('worst_case_max_dd_pct', 0):.1f} VP")
        print(f"Median Expected VP  : {data.get('median_ending_vp', 0):+.1f} VP")
        ci = data.get('confidence_interval_95', [0, 0])
        print(f"95% Confidence Int  : [{ci[0]:.1f}, {ci[1]:.1f}] VP")
        print(f"═════════════════════════════════════════════════════════════════════════")

def cmd_scorecard(args):
    data = request_api("scorecard", args.api_base)
    if args.export:
        export_json(data, args.export)
    elif args.json:
        print(json.dumps(data, indent=2))
    else:
        print("═════════════════════════════════════════════════════════════════════════")
        print("🏆 TRADERS FAMILY 7-PILLAR SCORECARD & REVENUE SHARE ELIGIBILITY")
        print("═════════════════════════════════════════════════════════════════════════")
        print(f"Total Score         : {data.get('total_score', 0)} / 28 ({data.get('total_score', 0)/28*100:.1f}%)")
        print(f"Revenue Share Tier  : {data.get('revenue_share_tier', 'LEGEND_PRIORITY')}")
        print(f"Max Revenue Share   : {data.get('max_revenue_share_pct', 80)}%")
        print("═════════════════════════════════════════════════════════════════════════")

def cmd_sync(args):
    symbol = args.symbol.upper()
    data = request_api(f"market/sync/{symbol}", args.api_base)
    if args.export:
        export_json(data, args.export)
    elif args.json:
        print(json.dumps(data, indent=2))
    else:
        print("═════════════════════════════════════════════════════════════════════════")
        print(f"⚡ DELTA SYNC REPORT: {symbol}")
        print("═════════════════════════════════════════════════════════════════════════")
        print(f"Source              : {data.get('source', 'DukascopyEcn')}")
        print(f"Synced Bars         : {data.get('synced_bars_count', 0)}")
        print(f"Duration            : {data.get('duration_ms', 0)} ms")
        print(f"Status              : {'🟢 100% Up-to-Date' if data.get('is_up_to_date') else '🔄 Synchronized'}")
        print(f"Message             : {data.get('message', '')}")
        print("═════════════════════════════════════════════════════════════════════════")

def cmd_dump_all(args):
    out_dir = args.out_dir or "reports/backtest"
    os.makedirs(out_dir, exist_ok=True)
    print(f"🚀 Memulai export seluruh data kuantitatif ke direktori: {out_dir}/ ...\n")
    
    # 1. Backtest Summary
    bt = request_api("backtest", args.api_base)
    export_json(bt, os.path.join(out_dir, "portfolio_backtest_summary.json"))
    
    # 2. Scorecard
    sc = request_api("scorecard", args.api_base)
    export_json(sc, os.path.join(out_dir, "scorecard_7_pillar.json"))
    
    # 3. Strategies
    strats = request_api("strategies", args.api_base)
    export_json(strats, os.path.join(out_dir, "strategies_catalog.json"))
    
    # 4. Per Pair Detailed, Trades, Monte Carlo, EDA
    pairs = ["EURGBP", "USDCHF", "GBPUSD", "EURUSD", "NZDUSD", "AUDUSD"]
    for p in pairs:
        print(f"\n📦 Mengambil dataset lengkap untuk pair: {p}...")
        try:
            det = request_api(f"backtest/detailed/{p}", args.api_base)
            export_json(det, os.path.join(out_dir, f"{p}_detailed_backtest.json"))
        except Exception as e:
            print(f"   ⚠️ Detailed backtest {p}: {e}")
            
        try:
            tr = request_api(f"backtest/trades/{p}", args.api_base)
            export_json(tr, os.path.join(out_dir, f"{p}_trades_list.json"))
        except Exception as e:
            print(f"   ⚠️ Trades {p}: {e}")
            
        try:
            mc = request_api(f"monte-carlo/{p}", args.api_base)
            export_json(mc, os.path.join(out_dir, f"{p}_monte_carlo.json"))
        except Exception as e:
            print(f"   ⚠️ Monte Carlo {p}: {e}")
            
        try:
            eda = request_api(f"eda/{p}", args.api_base)
            export_json(eda, os.path.join(out_dir, f"{p}_eda_health.json"))
        except Exception as e:
            print(f"   ⚠️ EDA {p}: {e}")

    print(f"\n🎉 SELESAI! Seluruh data JSON berhasil diekspor ke folder: {out_dir}/")

def main():
    parser = argparse.ArgumentParser(
        description="Hexagon Quant Terminal CLI & API Tool (2026)",
        formatter_class=argparse.RawTextHelpFormatter
    )
    parser.add_argument("--api-base", default=DEFAULT_API_BASE, help="REST API Base URL (default: http://127.0.0.1:5000/api)")
    subparsers = parser.add_subparsers(dest="command", help="Perintah yang tersedia")

    # health
    p_health = subparsers.add_parser("health", help="Cek status server API")
    p_health.add_argument("--json", action="store_true", help="Output raw JSON")
    p_health.add_argument("--export", metavar="PATH", help="Simpan hasil ke file JSON")

    # backtest
    p_bt = subparsers.add_parser("backtest", help="Ambil ringkasan portfolio backtest")
    p_bt.add_argument("--json", action="store_true", help="Output raw JSON")
    p_bt.add_argument("--export", metavar="PATH", help="Simpan hasil ke file JSON")

    # detailed
    p_det = subparsers.add_parser("detailed", help="Ambil detailed backtest report per pair")
    p_det.add_argument("symbol", help="Simbol pair (misal: EURGBP, EURUSD, NZDUSD)")
    p_det.add_argument("--json", action="store_true", help="Output raw JSON")
    p_det.add_argument("--export", metavar="PATH", help="Simpan hasil ke file JSON")

    # trades
    p_tr = subparsers.add_parser("trades", help="Ambil daftar settled trades per pair")
    p_tr.add_argument("symbol", help="Simbol pair (misal: EURGBP, EURUSD, NZDUSD)")
    p_tr.add_argument("--limit", type=int, help="Batasi jumlah trade yang ditampilkan di terminal")
    p_tr.add_argument("--json", action="store_true", help="Output raw JSON")
    p_tr.add_argument("--export", metavar="PATH", help="Simpan hasil ke file JSON")

    # monte-carlo
    p_mc = subparsers.add_parser("monte-carlo", help="Ambil laporan simulasi Monte Carlo 1.000 path")
    p_mc.add_argument("symbol", help="Simbol pair (misal: EURGBP, EURUSD, NZDUSD)")
    p_mc.add_argument("--json", action="store_true", help="Output raw JSON")
    p_mc.add_argument("--export", metavar="PATH", help="Simpan hasil ke file JSON")

    # scorecard
    p_sc = subparsers.add_parser("scorecard", help="Ambil skor 7-Pilar Traders Family")
    p_sc.add_argument("--json", action="store_true", help="Output raw JSON")
    p_sc.add_argument("--export", metavar="PATH", help="Simpan hasil ke file JSON")

    # sync
    p_sy = subparsers.add_parser("sync", help="Trigger atau cek delta sync status pair")
    p_sy.add_argument("symbol", help="Simbol pair (misal: EURGBP, EURUSD, NZDUSD)")
    p_sy.add_argument("--json", action="store_true", help="Output raw JSON")
    p_sy.add_argument("--export", metavar="PATH", help="Simpan hasil ke file JSON")

    # dump-all
    p_dump = subparsers.add_parser("dump-all", help="Ekspor seluruh data kuantitatif ke folder JSON")
    p_dump.add_argument("--out-dir", default="reports/backtest", help="Direktori output (default: reports/backtest)")

    args = parser.parse_args()
    if not args.command:
        parser.print_help()
        sys.exit(0)

    dispatch = {
        "health": cmd_health,
        "backtest": cmd_backtest,
        "detailed": cmd_detailed,
        "trades": cmd_trades,
        "monte-carlo": cmd_monte_carlo,
        "scorecard": cmd_scorecard,
        "sync": cmd_sync,
        "dump-all": cmd_dump_all,
    }

    if args.command in dispatch:
        dispatch[args.command](args)

if __name__ == "__main__":
    main()
